presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBuffer.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spark.execution;
+
+import com.facebook.presto.execution.buffer.OutputBufferMemoryManager;
+import com.facebook.presto.spark.classloader_interface.PrestoSparkRow;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+import static java.util.Objects.requireNonNull;
+
+public class PrestoSparkRowBuffer
+{
+    private final OutputBufferMemoryManager memoryManager;
+
+    private final Object monitor = new Object();
+    @GuardedBy("monitor")
+    private final Queue<PrestoSparkRow> buffer = new ArrayDeque<>();
+    @GuardedBy("monitor")
+    private volatile boolean finished;
+
+    public PrestoSparkRowBuffer(OutputBufferMemoryManager memoryManager)
+    {
+        this.memoryManager = requireNonNull(memoryManager, "memoryManager is null");
+    }
+
+    public ListenableFuture<?> isFull()
+    {
+        return memoryManager.getBufferBlockedFuture();
+    }
+
+    public void enqueue(PrestoSparkRow row)
+    {
+        requireNonNull(row, "row is null");
+        synchronized (monitor) {
+            buffer.add(row);
+            monitor.notify();
+        }
+        memoryManager.updateMemoryUsage(row.getRetainedSize());
+    }
+
+    public void setNoMoreRows()
+    {
+        memoryManager.setNoBlockOnFull();
+        synchronized (monitor) {
+            finished = true;
+            monitor.notifyAll();
+        }
+    }
+
+    public boolean hasRowsBuffered()
+    {
+        synchronized (monitor) {
+            return !buffer.isEmpty();
+        }
+    }
+
+    public PrestoSparkRow get()
+            throws InterruptedException
+    {
+        PrestoSparkRow row;
+        synchronized (monitor) {
+            while (!finished && buffer.isEmpty()) {
+                monitor.wait();
```

Comment:
So `PrestoSparkRowBuffer#get` is a blocked call. Is there any reason for that -- I am asking this since the general convention in Presto seems to be using `ListenableFuture<PrestoSparkRow>` for such cases?

Reply:
The Spark iterator is blocking anyway. It is mostly for implementation simplicity. Otherwise we would always wait on the `Future#get()` unconditionally in the SparkIterator.